module/zfs/dmu_recv.c

```diff
@@ -1207,6 +1257,261 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+
+	if (BP_USES_CRYPT(bp) && !rwa->raw) {
+		/*
+		 * If we're healing an encrypted block we should be using a
+		 * raw send or have the encyption keys loaded
+		 */
+		dsl_dataset_t *ds;
+		dsl_pool_t *dp = rwa->os->os_dsl_dataset->ds_dir->dd_pool;
+
+		/* get the dataset that the snap to be healed is in */
+		dsl_pool_config_enter(dp, FTAG);
+		/* load the key for this dataset */
+		err = dsl_dataset_hold_flags(dp, rwa->tofs,
+		    DS_HOLD_FLAG_DECRYPT, FTAG, &ds);
+		if (err != 0) {
+			dsl_pool_config_exit(dp, FTAG);
+			return (err);
+		}
+		err = spa_keystore_lookup_key(rwa->os->os_spa,
+		    dmu_objset_id(rwa->os), NULL, NULL);
+		dsl_dataset_rele_flags(ds, DS_HOLD_FLAG_DECRYPT, FTAG);
+		dsl_pool_config_exit(dp, FTAG);
+		if (err != 0)
+			return (SET_ERROR(EACCES));
+	}
+
+	/*
+	 * Raw healing is possible when the data is unchanged between the
+	 * send stream and on-disk
+	 */
+	if (rwa->raw) {
+		if (arc_is_encrypted(arc_buf) != BP_IS_ENCRYPTED(bp))
+			return (SET_ERROR(ENOTSUP));
+		if (arc_get_compression(arc_buf) != BP_GET_COMPRESS(bp))
+			return (SET_ERROR(ENOTSUP));
+		flags |= ZIO_FLAG_RAW;
+	}
+
+	/* Get the good data from the recv record */
+	abd = abd_get_from_buf(arc_buf->b_data, arc_buf_size(arc_buf));
+	abd_take_ownership_of_buf(abd, B_FALSE);
+
+	if ((arc_get_compression(arc_buf) != BP_GET_COMPRESS(bp)) ||
+	    (!arc_is_encrypted(arc_buf) && BP_USES_CRYPT(bp))) {
+		/*
+		 * The compression and/or encryption in the stream doesn't match
+		 * what we have on disk; we need to re-compress/re-encrypt the
+		 * buf to match what was written to disk previously.
+		 */
+		ASSERT(!rwa->raw);
+		if (arc_get_compression(arc_buf) != ZIO_COMPRESS_OFF) {
+			/* Decompress the stream data */
+			err = zio_decompress_data(arc_get_compression(arc_buf),
+			    abd, buf, arc_buf_size(arc_buf), lsize);
+			abd_release_ownership_of_buf(abd);
+			abd_put(abd);
+			if (err != 0)
+				return (err);
+			/* Swap in the newly decompressed data into the abd */
+			abd = abd_get_from_buf(buf, lsize);
+			abd_take_ownership_of_buf(abd, B_FALSE);
+		}
+
+		/*
+		 * The stream is not encrypted but the data on-disk is.
+		 * We need to re-encrypt the buf using the same
+		 * encryption type, salt, iv, and mac that was used to encrypt
+		 * the block previosly.
+		 */
+		if (!arc_is_encrypted(arc_buf) && BP_USES_CRYPT(bp)) {
+			dsl_dataset_t *ds;
+			dsl_crypto_key_t *dck = NULL;
+			uint8_t salt[ZIO_DATA_SALT_LEN];
+			uint8_t iv[ZIO_DATA_IV_LEN];
+			uint8_t mac[ZIO_DATA_MAC_LEN];
+			boolean_t no_crypt = B_FALSE;
+			dsl_pool_t *dp =
+			    rwa->os->os_dsl_dataset->ds_dir->dd_pool;
+			enc_buf = zio_data_buf_alloc(lsize);
+			abd_t *eabd = abd_get_from_buf(enc_buf, lsize);
+			abd_take_ownership_of_buf(eabd, B_FALSE);
+			zio_crypt_decode_params_bp(bp, salt, iv);
+			zio_crypt_decode_mac_bp(bp, mac);
+
+			dsl_pool_config_enter(dp, FTAG);
+			err = dsl_dataset_hold_flags(dp, rwa->tofs,
+			    DS_HOLD_FLAG_DECRYPT, FTAG, &ds);
```

Comment:
`dsl_dataset_hold_flags(..., DS_HOLD_FLAG_DECRYPT, ...)` ensures that the keymap is created so that we can look it up later on with the `spa_keystore_lookup_key() call`